### PR TITLE
Bst example

### PR DESCRIPTION
**Pull request by harrisonbrock**
This is an WIP (work in progress)

### CODE REVIEW

File: topics/go/algorithms/data/bst/bst.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+		Data:  value,
+		left:  nil,
+		right: nil,
+	}
+
+	// If the root node is nil set the root node
+	// to n
+	if bst.root == nil {
+		bst.root = n
+	} else {
```
Comment (ardan-bkennedy) [inline topics/go/algorithms/data/bst/bst.go:40]: Get rid of the else clause.

File: topics/go/algorithms/data/bst/bst.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+		root: nil,
+	}
+}
+
+// Insert an integer into the BST.
+func (bst *BST) Insert(value int) {
+
+	// Create a new node.
+	n := &Node{
+		Data:  value,
```
Comment (ardan-bkennedy) [inline topics/go/algorithms/data/bst/bst.go:31]: Don't use pointer semantic contruction to a variable.

File: topics/go/algorithms/data/bst/bst.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+	// to n
+	if bst.root == nil {
+		bst.root = n
+	} else {
+		insertNode(bst.root, n)
+	}
+}
+
+// Max returns the max int
+func (bst *BST) Max() (int, error) {
```
Comment (ardan-bkennedy) [inline topics/go/algorithms/data/bst/bst.go:46]: Check all comments. comments can be extended within the scope of a function (I'm looking above).

File: topics/go/algorithms/data/bst/bst.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+	currentNode := bst.root
+
+	if currentNode == nil {
+		return 0, fmt.Errorf("root node is nil")
+	}
+
+	for {
+		if currentNode.right == nil {
```
Comment (ardan-bkennedy) [inline topics/go/algorithms/data/bst/bst.go:55]: This loops until the right side is nil. Wondering if that can be incorporated into the for.  I could be wrong.

File: topics/go/algorithms/data/bst/bst.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	if newNode.Data < root.Data {
+		if root.left == nil {
+			root.left = newNode
+		} else {
+			insertNode(root.left, newNode)
+		}
+
+		// Insert into the right side of the tree.
+	} else {
+		if root.right == nil {
```
Comment (ardan-bkennedy) [inline topics/go/algorithms/data/bst/bst.go:107]: Stay away from else.  You can return early or use a naked switch

File: topics/go/algorithms/data/bst/bst.go [hunk 6]
```diff
@@ -0,0 +1,113 @@
+package bst
```
Comment (ardan-bkennedy) [inline topics/go/algorithms/data/bst/bst.go:1]: We need some information about what a binary search tree is.  Add a doc.go and provide some information.

File: topics/go/algorithms/data/bst/bst.go [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+// BST represents a binary search tree.
+type BST struct {
+	root *Node
+}
+
+// New creates a new BST.
+func New() *BST {
+	return &BST{
+		root: nil,
```
Comment (ardan-bkennedy) [inline topics/go/algorithms/data/bst/bst.go:22]: Don't need to specify zero value fields. So this is usable in its zero value state. This means let's remove the factor function. A use can do this.  var bst BST

File: topics/go/algorithms/data/bst/bst.go (hunk 2 above)
Comment (balta2ar) [reply topics/go/algorithms/data/bst/bst.go:31]: Why?

File: topics/go/algorithms/data/bst/bst.go (hunk 3 above)
Comment (harrisonbrock) [reply topics/go/algorithms/data/bst/bst.go:46]: Well, do.

File: topics/go/algorithms/data/bst/bst.go (hunk 2 above)
Comment (ardan-bkennedy) [reply topics/go/algorithms/data/bst/bst.go:31]: For readability. The & shows sharing and you don't want to lose that information.

We make things easy to understand, not easy to do.

**Pull request closed without merging**